backend/regalloc/regalloc_split_state.ml

```diff
@@ -43,6 +43,176 @@ let log_renaming_info : indent:int -> t -> unit =
+              then
+                let to_move : Reg.Set.t =
+                  Reg.Set.filter
+                    (fun (reg : Reg.t) ->
+                      let occurs = occurs_block block reg in
+                      not occurs)
+                    definitions
+                in
+                if not (Reg.Set.is_empty to_move)
+                then (
+                  if split_debug
+                  then
+                    log ~indent:2 "moving %a from block %d to block %d"
+                      Printmach.regset to_move label successor_label;
+                  definitions_at_beginning
+                    := Label.Map.update label
+                         (function
+                           | None -> assert false
+                           | Some set -> Some (Reg.Set.diff set to_move))
+                         !definitions_at_beginning;
+                  definitions_at_beginning
+                    := Label.Map.update successor_label
+                         (function
+                           | None -> Some to_move
+                           | Some set -> Some (Reg.Set.union set to_move))
+                         !definitions_at_beginning)
+            | _ -> ()));
+    !definitions_at_beginning, stack
+
+  (** A destruction at the end of a block can be move up if:
+
+      - the block is not a trap handler;
+      - the block has only one predecessor;
+      - the predecessor has only one (normal) successor;
+      - the predecessor does neither raise nor end with a destruction point;
+      - the block and the predecessor are different;
+      - the destroyed register has no occurrences in the block
+        (including definitions). *)
+  let move_destructions_at_end_up :
+      Cfg_with_infos.t ->
+      Label.t Stack.t ->
+      definitions_at_beginning:definitions_at_beginning ->
+      destructions_at_end:destructions_at_end ->
+      destructions_at_end =
+   fun cfg_with_infos stack ~definitions_at_beginning ~destructions_at_end ->
+    if split_debug
+    then log ~indent:1 "MoveSpillsAndReloads.move_destructions_at_end_up";
+    let destructions_at_end = ref destructions_at_end in
+    while not (Stack.is_empty stack) do
+      let label = Stack.pop stack in
+      let block = Cfg_with_infos.get_block_exn cfg_with_infos label in
+      match Label.Map.find_opt label !destructions_at_end with
+      | None -> ()
+      | Some (destruction_kind, destructions) ->
+        if (not block.is_trap_handler)
+           && Label.Set.cardinal block.predecessors = 1
+        then
+          let predecessor_label = Label.Set.choose block.predecessors in
+          let predecessor_block =
+            Cfg.get_block_exn
+              (Cfg_with_infos.cfg cfg_with_infos)
+              predecessor_label
+          in
+          if (not (Label.equal label predecessor_label))
+             && Option.is_none predecessor_block.exn
+             && Option.is_none (destruction_point_at_end predecessor_block)
```

Comment:
why is this condition needed?

Reply:
I think it is (at least currently) partially
redundant, but the idea is that spills and
reloads should not cross destruction
points.